### ISSUE TITLE
[Tabs][Carousel][Accordion][Accessibility] Tabs don't have aria-controls and tabpanels don't have aria-labeledby

### ISSUE THREAD

**Issue by jakublamprecht** [adobe/aem-core-wcm-components#588]
## Bug Report

**Current Behavior**
Current markup of Tabs, Accordion and Carousel components does not tell the user about the relationship between tabs and corresponding tabpanels. To achieve that tabs should have an `aria-controls` attribute which value equals to corresponding tabpanels id, and tabpanels should have an `aria-labeledby` attribute which value equals to corresponding tabs id.

More information on this can be found here:
https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel
https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html
https://davatron5000.github.io/a11y-nutrition-cards/components/tabs

**Expected behavior/code**
Markup has `aria-controls` and `aria-labeledby` attributes on respectively tab and tabpanel.

**Environment**
- Core Components version 2.3.0

**Comment by A30001907** [comment]
In addition to the 'aria'- attributes above, the currently open/active tab should have aria-expanded="true" attribute added as that tab is activated/opened, and aria-expanded="false" as the tab is de-activate/closed.

**Comment by jakublamprecht** [comment]
@A30001907 I believe this is not necessary, as `aria-selected` attribute is added with JS on currently active tab. As suggested by W3C, tabs component should use `aria-selected` and accordion component should use `aria-expanded`.

**Comment by A30001907** [comment]
@jakublamprecht it is possible to use the keyboard to navigate the tabs via the tab key or with a screen reader activated by action keys, as such aria-selected is announced on a tab that is not active, but does however have keyboard focus, thus it is 'selected', but remains 'collapsed' and should be announced as such. Try this out yourself by using the tab key to navigate through the tab component tabs and observe the screen reader will announce aria-selected on each tab even though it is not physically the active tab. Hit enter key or space bar to activate the tab and note that it is not announced as expanded nor selected.

**Comment by jakublamprecht** [comment]
@A30001907 As far as W3C goes, using the tab key when focused on a `tab` element should move the focus to the `tabpanel` element and not the next `tab` element. Looking at the code, I can see that only the active tab has `tabindex=0` and other `tabs` are not tabbable, so I guess it should work correctly. I tabbed through the demo page of Tabs component with screen reader on and everything seemed correct

**Comment by A30001907** [comment]
@jakublamprecht using the tab key should not focus into the _tabpanel_, it should move to the next _tab_ to give an assistive technology user an opportunity to browse across the tabs to decide whether they would like to open them. Additionally, keyboard navigating via tab keys will only locate focusable elements, a _tabpanel_ should not be focusable. Setting of _tabindex_ should be avoided where-ever possible, by the very nature of a screen readers linear approach to reading a web page, top-down-left-to-right, setting of _tabindex_ will likely interfere with this and cause users to miss content. The exception to this rule would be setting tabIndex='-1' on an element that is otherwise not keyboard focusable, but needs to receive focus after some action in page. eg: an action has modified content, so the h1/2/3/4 in the new content should receive focus.

**Comment by richardhand** [comment]
@A30001907 we base our tabs implementation on [WAI-ARIA 1.1](https://www.w3.org/TR/wai-aria-practices), specifically, the [tabs design pattern](https://www.w3.org/TR/wai-aria-practices/#tabpanel). In that pattern `tablist`s are not navigated with a tab key, rather with left/right arrows.

Also see the [tabs with automatic activation](https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html) example which closely follows the behaviour of the tabs core component. 

This issue is about filling in the gaps that are missing from the specification. I'm also still not sure about blindly setting the `tabindex` for all tabpanels without some check to ensure there is no focusable element in the panel, as it creates an additional stop when tabbing through the document. Although, it _is_ listed in the WAI-ARIA tabs example.

**Comment by jakublamprecht** [comment]
@A30001907 Wanted to say exactly what @richardhand said. I don't consider myself an expert in accessibility by any means - everything I've said was based on the links that @richardhand provided.

And how about we move this discussion to a separate issue if that's still an issue, as this is not directly related to the missing `aria-controls` and `aria-labeledby` attributes?

**Comment by jakublamprecht** [comment]
I forgot to mention it in the ticket, but the `aria-controls` and `aria-labeledby` are also missing for Carousel and Accordion components (the main issue with that is that elements do not have unique id's)

Just realised, that except the things i've mentioned in the issue, tablist in tabs component is also missing aria-label as specified here:

https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel

> "If the tab list has a visible label, the element with role tablist has aria-labelledby set to a value that refers to the labeling element. Otherwise, the tablist element has a label provided by aria-label."

The same goes for Carousel component which uses tab carousel pattern and also has element with role 'tablist' but no "aria-label".

@richardhand would you rather have me edit this ticket or create separate issues for those findings?

**Comment by richardhand** [comment]
@jakublamprecht - would be great if you could create a separate issue for the other missing `aria-label`s, the `aria-controls` and `aria-labeledby` for Carousel and Accordion can also be tackled in this issue, we should probably just update the title and description here.

**Comment by jakublamprecht** [comment]
@richardhand Darn you're fast :) One more question - do you have any plans as to when you would want to adress this controls/labeledby issue?

**Comment by richardhand** [comment]
This issue is targeted for the next release, but we don't yet have a set time frame for when that will be available.

**Comment by jakublamprecht** [comment]
Cool, thanks for the information :)
